File: python/prism_sdk/autonomous_task_decision.py
"""Provider-free intent-to-action posture for the autonomous brain.

Task intent makes the first interpretation inspectable.  This module takes that bounded
interpretation one step further: it derives the next safe execution posture from the reviewed
intent, domain policy, and task lens.  The result is guidance and admission metadata, never an
authorization token.  Provider, source, tool, credential, effect, and evaluator boundaries stay
owned by their existing callers.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping, Sequence

from .authoring import content_digest
from .autonomous_domain_policy import AutonomousDomainPolicy
from .autonomous_task_intent import (
    AUTONOMOUS_TASK_INTENT_DOMAINS,
    AUTONOMOUS_TASK_INTENT_EFFECTS,
    AutonomousTaskIntent,
)
from .autonomous_task_lens import AutonomousDomainTaskLens
from .errors import ArgumentError


AUTONOMOUS_TASK_DECISION_SCHEMA = "bioprism-autonomous-task-decision/0.1"
AUTONOMOUS_TASK_DECISION_VERSION = "0.1"
AUTONOMOUS_TASK_DECISION_POSTURES = ("admitted", "review_required", "blocked")
AUTONOMOUS_TASK_DECISION_PATHS = (
    "provider",
    "evidence_first",
    "workflow",
    "planning",
    "cross_domain",
)
AUTONOMOUS_TASK_DECISION_APPROVALS = (
    "provider_call",
    "evidence_dispatch",
    "plan_acceptance",
    "effect_approval",
    "evaluator_settlement",
)
AUTONOMOUS_TASK_DECISION_EVIDENCE_POSTURES = ("optional", "required_before_provider")
MAX_AUTONOMOUS_TASK_DECISION_ITEMS = 12
MAX_AUTONOMOUS_TASK_DECISION_TEXT_BYTES = 512


def _bounded_text(name: str, value: Any) -> str:
    if not isinstance(value, str) or not value.strip() or "\x00" in value:
        raise ArgumentError(f"{name} must be a non-empty string")
    if len(value.encode("utf-8")) > MAX_AUTONOMOUS_TASK_DECISION_TEXT_BYTES:
        raise ArgumentError(f"{name} exceeds its bound")
    return value


def _bounded_items(name: str, values: Sequence[str]) -> tuple[str, ...]:
    if not isinstance(values, Sequence) or isinstance(values, (str, bytes)) or len(values) > MAX_AUTONOMOUS_TASK_DECISION_ITEMS:
        raise ArgumentError(f"{name} exceeds its item bound")
    result = tuple(_bounded_text(f"{name} item", value) for value in values)
    if len(set(result)) != len(result):
        raise ArgumentError(f"{name} contains duplicate items")
    return result


def _digest(name: str, value: str) -> str:
    if not isinstance(value, str) or len(value) != 64 or any(character not in "0123456789abcdef" for character in value):
        raise ArgumentError(f"{name} must be a lowercase SHA-256 digest")
    return value


def _unique(values: Sequence[str]) -> tuple[str, ...]:
    return tuple(dict.fromkeys(values))


@dataclass(frozen=True, slots=True)
class AutonomousTaskDecision:
    """A digest-bound recommendation for the next autonomous execution boundary."""

    domain: str
    workflow_id: str
    task_digest: str
    intent_id: str
    intent_digest: str
    lens_digest: str
    policy_digest: str
    decision_id: str
    posture: str
    recommended_path: str
    requested_effect: str
    evidence_posture: str
    required_model_capabilities: tuple[str, ...]
    preferred_model_capabilities: tuple[str, ...]
    approval_requirements: tuple[str, ...]
    review_reasons: tuple[str, ...]
    blocking_reasons: tuple[str, ...]
    next_actions: tuple[str, ...]
    decision_version: str = AUTONOMOUS_TASK_DECISION_VERSION

    def __post_init__(self) -> None:
        if self.domain not in AUTONOMOUS_TASK_INTENT_DOMAINS:
            raise ArgumentError(f"unsupported task-decision domain: {self.domain}")
        for name, value in (
            ("domain", self.domain),
            ("workflow_id", self.workflow_id),
            ("intent_id", self.intent_id),
            ("decision_id", self.decision_id),
        ):
            _bounded_text(f"task decision {name}", value)
        if self.decision_version != AUTONOMOUS_TASK_DECISION_VERSION:
            raise ArgumentError("unsupported task-decision version")
        for name, value in (
            ("task_digest", self.task_digest),
            ("intent_digest", self.intent_digest),
            ("lens_digest", self.lens_digest),
            ("policy_digest", self.policy_digest),
        ):
            _digest(f"task decision {name}", value)
        if self.posture not in AUTONOMOUS_TASK_DECISION_POSTURES:
            raise ArgumentError("task decision posture is unsupported")
        if self.recommended_path not in AUTONOMOUS_TASK_DECISION_PATHS:
            raise ArgumentError("task decision recommended_path is unsupported")
        if self.requested_effect not in AUTONOMOUS_TASK_INTENT_EFFECTS:
            raise ArgumentError("task decision requested_effect is unsupported")
        if self.evidence_posture not in AUTONOMOUS_TASK_DECISION_EVIDENCE_POSTURES:
            raise ArgumentError("task decision evidence_posture is unsupported")
        for name, value in (
            ("required_model_capabilities", self.required_model_capabilities),
            ("preferred_model_capabilities", self.preferred_model_capabilities),
            ("approval_requirements", self.approval_requirements),
            ("review_reasons", self.review_reasons),
            ("blocking_reasons", self.blocking_reasons),
            ("next_actions", self.next_actions),
        ):
            _bounded_items(f"task decision {name}", value)
        if any(value not in AUTONOMOUS_TASK_DECISION_APPROVALS for value in self.approval_requirements):
            raise ArgumentError("task decision approval_requirements contains an unsupported gate")

    def _descriptor(self) -> dict[str, Any]:
        return {
            "schema": AUTONOMOUS_TASK_DECISION_SCHEMA,
            "decision_version": self.decision_version,
            "domain": self.domain,
            "workflow_id": self.workflow_id,
            "task_digest": self.task_digest,
            "intent_id": self.intent_id,
            "intent_digest": self.intent_digest,
            "lens_digest": self.lens_digest,
            "policy_digest": self.policy_digest,
            "decision_id": self.decision_id,
            "posture": self.posture,
            "recommended_path": self.recommended_path,
            "requested_effect": self.requested_effect,
            "evidence_posture": self.evidence_posture,
            "required_model_capabilities": list(self.required_model_capabilities),
            "preferred_model_capabilities": list(self.preferred_model_capabilities),
            "approval_requirements": list(self.approval_requirements),
            "review_reasons": list(self.review_reasons),
            "blocking_reasons": list(self.blocking_reasons),
            "next_actions": list(self.next_actions),
        }

    @property
    def decision_digest(self) -> str:
        return content_digest(self._descriptor())

    def to_dict(self) -> dict[str, Any]:
        return {
            **self._descriptor(),
            "decision_digest": self.decision_digest,
            "authorization": "guidance_only;provider_source_tool_and_effect_authority_remain_separate",
            "retention": "value_only_decision_metadata;task_text_not_retained",
            "secret_material": "never_returned",
        }

    def prompt_contract(self, *, compact: bool = False) -> dict[str, Any]:
        result: dict[str, Any] = {
            "schema": AUTONOMOUS_TASK_DECISION_SCHEMA,
            "decision_id": self.decision_id,
            "decision_digest": self.decision_digest,
            "intent_digest": self.intent_digest,
            "posture": self.posture,
            "recommended_path": self.recommended_path,
            "requested_effect": self.requested_effect,
            "evidence_posture": self.evidence_posture,
            "approval_requirements": list(self.approval_requirements),
            "review_reasons": list(self.review_reasons),
            "blocking_reasons": list(self.blocking_reasons),
            "authority": "guidance_only;does_not_authorize_provider_source_tool_or_effect_actions",
        }
        if not compact:
            result.update(
                {
                    "required_model_capabilities": list(self.required_model_capabilities),
                    "preferred_model_capabilities": list(self.preferred_model_capabilities),
                    "next_actions": list(self.next_actions),
                }
            )
        result["secret_material"] = "never_returned"
        return result


def infer_autonomous_task_decision(
    *,
    intent: AutonomousTaskIntent,
    lens: AutonomousDomainTaskLens,
    policy: AutonomousDomainPolicy,
    required_model_capabilities: Sequence[str] = (),
) -> AutonomousTaskDecision:
    """Derive an auditable next-step posture without invoking a provider or tool."""

    if not isinstance(intent, AutonomousTaskIntent) or not isinstance(lens, AutonomousDomainTaskLens) or not isinstance(policy, AutonomousDomainPolicy):
        raise ArgumentError("task decision requires a valid intent, lens, and policy")
    if intent.domain != lens.domain or intent.domain != policy.domain:
        raise ArgumentError("task decision intent, lens, and policy domains must agree")
    required = _bounded_items("task decision required_model_capabilities", required_model_capabilities)
    if not required:
        raise ArgumentError("task decision requires at least one model capability")

    action = intent.action_mode
    if intent.domain == "cross_domain" or action in {"coordinate", "synthesize"}:
        path = "cross_domain"
    elif action in {"create", "modify"}:
        path = "workflow"
    elif action in {"plan"}:
        path = "planning"
    elif policy.evidence_mode == "required_before_provider" or action in {"observe", "investigate", "analyze", "compare", "evaluate"}:
        path = "evidence_first"
    else:
        path = "provider"

    preferred = list(lens.model_capability_hints)
    if path == "workflow":
        preferred.append("structured_output")
    if path == "cross_domain":
        preferred.extend(("coordination", "structured_output"))
    if action in {"analyze", "compare", "evaluate", "synthesize"}:
        preferred.append("reasoning")
    preferred = list(_unique(preferred))

    approvals = ["provider_call"]
    review: list[str] = []
    blocked: list[str] = []
    if policy.evidence_mode == "required_before_provider":
        approvals.append("evidence_dispatch")
        review.append("evidence_required_before_provider")
    if path in {"workflow", "planning", "cross_domain"} and policy.plan_acceptance_required:
        approvals.append("plan_acceptance")
        review.append("plan_acceptance_required")
    if intent.requested_effect == "external_effect":
        if policy.effect_mode == "forbidden":
            blocked.append("requested_effect_forbidden_by_domain_policy")
        else:
            approvals.append("effect_approval")
            review.append("external_effect_requires_explicit_approval")
    elif intent.requested_effect == "local_change" and policy.effect_mode == "approval_gated":
        approvals.append("effect_approval")
        review.append("local_change_requires_explicit_approval")
    if policy.evaluator_required:
        approvals.append("evaluator_settlement")
    for flag in intent.ambiguity_flags:
        review.append(f"intent:{flag}")
    if action in {"coordinate", "synthesize"} or intent.domain == "cross_domain":
        review.append("specialist_boundaries_require_review")
    if intent.risk_signals:
        review.append("domain_risk_signals_require_review")

    approvals = list(_unique(approvals))
    review = list(_unique(review))
    blocked = list(_unique(blocked))
    posture = "blocked" if blocked else "review_required" if review else "admitted"
    if blocked:
        next_actions = ["stop_before_provider_dispatch", "resolve_domain_policy_conflict"]
    elif review:
        next_actions = ["review_task_intent_and_decision", "satisfy_required_approval_gates"]
        if "evidence_dispatch" in approvals:
            next_actions.append("acquire_and_review_required_evidence")
    else:
        next_actions = ["select_model", "request_provider_call_approval"]
    if "evaluator_settlement" in approvals:
        next_actions.append("settle_explicit_evaluator_feedback_after_run")
    next_actions = list(_unique(next_actions))
    return AutonomousTaskDecision(
        domain=intent.domain,
        workflow_id=intent.workflow_id,
        task_digest=intent.task_digest,
        intent_id=intent.intent_id,
        intent_digest=intent.intent_digest,
        lens_digest=lens.lens_digest,
        policy_digest=policy.policy_digest,
        decision_id=f"{intent.intent_id}:{posture}:{path}",
        posture=posture,
        recommended_path=path,
        requested_effect=intent.requested_effect,
        evidence_posture=policy.evidence_mode,
        required_model_capabilities=required,
        preferred_model_capabilities=tuple(preferred),
        approval_requirements=tuple(approvals),
        review_reasons=tuple(review),
        blocking_reasons=tuple(blocked),
        next_actions=tuple(next_actions),
    )


def validate_autonomous_task_decision(
    value: AutonomousTaskDecision | Mapping[str, Any],
    *,
    intent: AutonomousTaskIntent | None = None,
    lens: AutonomousDomainTaskLens | None = None,
    policy: AutonomousDomainPolicy | None = None,
    required_model_capabilities: Sequence[str] | None = None,
) -> AutonomousTaskDecision:
    """Validate a persisted decision and optionally replay it against live task artifacts.

    A serialized decision is guidance metadata, not an authorization token.  Structural
    validation checks its digest, markers, bounded fields, and enum values.  When the original
    intent, lens, and policy are available, the deterministic decision is recomputed and every
    descriptor field must match.  This prevents stale or tampered decisions from crossing a
    restart boundary into provider, source, tool, evaluator, or effect execution.
    """

    if isinstance(value, AutonomousTaskDecision):
        decision = value
    else:
        if not isinstance(value, Mapping):
            raise ArgumentError("task decision must be an object")
        allowed = {
            "schema", "decision_version", "domain", "workflow_id", "task_digest", "intent_id",
            "intent_digest", "lens_digest", "policy_digest", "decision_id", "posture",
            "recommended_path", "requested_effect", "evidence_posture",
            "required_model_capabilities", "preferred_model_capabilities", "approval_requirements",
            "review_reasons", "blocking_reasons", "next_actions", "decision_digest",
            "authorization", "retention", "secret_material",
        }
        if set(value).difference(allowed):
            raise ArgumentError("task decision contains unsupported fields")
        if (
            value.get("schema") != AUTONOMOUS_TASK_DECISION_SCHEMA
            or value.get("authorization") != "guidance_only;provider_source_tool_and_effect_authority_remain_separate"
            or value.get("retention") != "value_only_decision_metadata;task_text_not_retained"
            or value.get("secret_material") != "never_returned"
        ):
            raise ArgumentError("task decision markers are invalid")

        def items(name: str) -> tuple[str, ...]:
            raw = value.get(name)
            if not isinstance(raw, Sequence) or isinstance(raw, (str, bytes)):
                raise ArgumentError(f"task decision {name} must be a sequence")
            return tuple(raw)

        try:
            decision = AutonomousTaskDecision(
                domain=value.get("domain"),
                workflow_id=value.get("workflow_id"),
                task_digest=value.get("task_digest"),
                intent_id=value.get("intent_id"),
                intent_digest=value.get("intent_digest"),
                lens_digest=value.get("lens_digest"),
                policy_digest=value.get("policy_digest"),
                decision_id=value.get("decision_id"),
                posture=value.get("posture"),
                recommended_path=value.get("recommended_path"),
                requested_effect=value.get("requested_effect"),
                evidence_posture=value.get("evidence_posture"),
                required_model_capabilities=items("required_model_capabilities"),
                preferred_model_capabilities=items("preferred_model_capabilities"),
                approval_requirements=items("approval_requirements"),
                review_reasons=items("review_reasons"),
                blocking_reasons=items("blocking_reasons"),
                next_actions=items("next_actions"),
                decision_version=value.get("decision_version"),
            )
        except (TypeError, ValueError) as error:
            raise ArgumentError("task decision fields are malformed") from error
        if value.get("decision_digest") != decision.decision_digest:
            raise ArgumentError("task decision digest does not match its metadata")

    bindings = (intent, lens, policy)
    if any(item is not None for item in bindings):
        if not all(isinstance(item, expected) for item, expected in zip(bindings, (AutonomousTaskIntent, AutonomousDomainTaskLens, AutonomousDomainPolicy))):
            raise ArgumentError("task decision replay requires intent, lens, and policy together")
        replay = infer_autonomous_task_decision(
            intent=intent,
            lens=lens,
            policy=policy,
            required_model_capabilities=(
                decision.required_model_capabilities
                if required_model_capabilities is None
                else required_model_capabilities
            ),
        )
        if replay._descriptor() != decision._descriptor():
            raise ArgumentError("task decision does not match the supplied intent, lens, and policy")
    elif required_model_capabilities is not None:
        raise ArgumentError("task decision replay capabilities require intent, lens, and policy")
    return decision


__all__ = [
    "AUTONOMOUS_TASK_DECISION_SCHEMA",
    "AUTONOMOUS_TASK_DECISION_VERSION",
    "AUTONOMOUS_TASK_DECISION_POSTURES",
    "AUTONOMOUS_TASK_DECISION_PATHS",
    "AUTONOMOUS_TASK_DECISION_APPROVALS",
    "AUTONOMOUS_TASK_DECISION_EVIDENCE_POSTURES",
    "MAX_AUTONOMOUS_TASK_DECISION_ITEMS",
    "AutonomousTaskDecision",
    "infer_autonomous_task_decision",
    "validate_autonomous_task_decision",
]

File: typescript/src/autonomous-task-decision.ts
import { ArgumentError } from "./errors.js";
import { validateAutonomousDomainPolicy, type AutonomousDomainPolicy } from "./autonomous-domain-policy.js";
import { validateAutonomousDomainTaskLens, type AutonomousDomainTaskLens } from "./autonomous-task-lens.js";
import { AUTONOMOUS_TASK_INTENT_DOMAINS, validateAutonomousTaskIntent, type AutonomousTaskIntent } from "./autonomous-task-intent.js";
import { digestJsonSync } from "./tooling.js";
import type { JsonObject } from "./types.js";

export const AUTONOMOUS_TASK_DECISION_SCHEMA = "bioprism-autonomous-task-decision/0.1" as const;
export const AUTONOMOUS_TASK_DECISION_VERSION = "0.1" as const;
export const AUTONOMOUS_TASK_DECISION_POSTURES = ["admitted", "review_required", "blocked"] as const;
export const AUTONOMOUS_TASK_DECISION_PATHS = ["provider", "evidence_first", "workflow", "planning", "cross_domain"] as const;
export const AUTONOMOUS_TASK_DECISION_APPROVALS = ["provider_call", "evidence_dispatch", "plan_acceptance", "effect_approval", "evaluator_settlement"] as const;
export const AUTONOMOUS_TASK_DECISION_EVIDENCE_POSTURES = ["optional", "required_before_provider"] as const;
export const MAX_AUTONOMOUS_TASK_DECISION_ITEMS = 12;
const MAX_DECISION_TEXT_BYTES = 512;

function text(name: string, value: unknown): string {
  if (typeof value !== "string" || !value.trim() || value.includes("\u0000") || new TextEncoder().encode(value).byteLength > MAX_DECISION_TEXT_BYTES) throw new ArgumentError(`${name} is outside the task-decision text bound`);
  return value;
}

function items(name: string, value: unknown): string[] {
  if (!Array.isArray(value) || value.length > MAX_AUTONOMOUS_TASK_DECISION_ITEMS) throw new ArgumentError(`${name} exceeds the task-decision item bound`);
  const result = value.map((item) => text(`${name} item`, item));
  if (new Set(result).size !== result.length) throw new ArgumentError(`${name} contains duplicate items`);
  return result;
}

function digest(name: string, value: unknown): string {
  if (typeof value !== "string" || !/^[0-9a-f]{64}$/.test(value)) throw new ArgumentError(`${name} must be a lowercase SHA-256 digest`);
  return value;
}

function unique(values: readonly string[]): string[] { return [...new Set(values)]; }

export interface AutonomousTaskDecision extends JsonObject {
  schema: typeof AUTONOMOUS_TASK_DECISION_SCHEMA;
  decision_version: typeof AUTONOMOUS_TASK_DECISION_VERSION;
  domain: AutonomousTaskIntent["domain"];
  workflow_id: string;
  task_digest: string;
  intent_id: string;
  intent_digest: string;
  lens_digest: string;
  policy_digest: string;
  decision_id: string;
  posture: typeof AUTONOMOUS_TASK_DECISION_POSTURES[number];
  recommended_path: typeof AUTONOMOUS_TASK_DECISION_PATHS[number];
  requested_effect: AutonomousTaskIntent["requested_effect"];
  evidence_posture: AutonomousDomainPolicy["evidence_mode"];
  required_model_capabilities: string[];
  preferred_model_capabilities: string[];
  approval_requirements: string[];
  review_reasons: string[];
  blocking_reasons: string[];
  next_actions: string[];
  decision_digest: string;
  authorization: "guidance_only;provider_source_tool_and_effect_authority_remain_separate";
  retention: "value_only_decision_metadata;task_text_not_retained";
  secret_material: "never_returned";
}

type DecisionDescriptor = Omit<AutonomousTaskDecision, "decision_digest" | "authorization" | "retention" | "secret_material">;

function descriptorFor(decision: AutonomousTaskDecision | DecisionDescriptor): DecisionDescriptor {
  const value = decision as AutonomousTaskDecision;
  return {
    schema: AUTONOMOUS_TASK_DECISION_SCHEMA,
    decision_version: AUTONOMOUS_TASK_DECISION_VERSION,
    domain: value.domain,
    workflow_id: value.workflow_id,
    task_digest: value.task_digest,
    intent_id: value.intent_id,
    intent_digest: value.intent_digest,
    lens_digest: value.lens_digest,
    policy_digest: value.policy_digest,
    decision_id: value.decision_id,
    posture: value.posture,
    recommended_path: value.recommended_path,
    requested_effect: value.requested_effect,
    evidence_posture: value.evidence_posture,
    required_model_capabilities: [...value.required_model_capabilities],
    preferred_model_capabilities: [...value.preferred_model_capabilities],
    approval_requirements: [...value.approval_requirements],
    review_reasons: [...value.review_reasons],
    blocking_reasons: [...value.blocking_reasons],
    next_actions: [...value.next_actions],
  };
}

export function autonomousTaskDecisionDigest(decision: AutonomousTaskDecision | DecisionDescriptor): string {
  return digestJsonSync(descriptorFor(decision));
}

export function autonomousTaskDecisionPromptContract(decision: AutonomousTaskDecision, compact = false): JsonObject {
  if (!decision || decision.schema !== AUTONOMOUS_TASK_DECISION_SCHEMA) throw new ArgumentError("task decision prompt contract requires a valid decision");
  const result: JsonObject = {
    schema: AUTONOMOUS_TASK_DECISION_SCHEMA,
    decision_id: decision.decision_id,
    decision_digest: decision.decision_digest,
    intent_digest: decision.intent_digest,
    posture: decision.posture,
    recommended_path: decision.recommended_path,
    requested_effect: decision.requested_effect,
    evidence_posture: decision.evidence_posture,
    approval_requirements: [...decision.approval_requirements],
    review_reasons: [...decision.review_reasons],
    blocking_reasons: [...decision.blocking_reasons],
    authority: "guidance_only;does_not_authorize_provider_source_tool_or_effect_actions",
  };
  if (!compact) Object.assign(result, { required_model_capabilities: [...decision.required_model_capabilities], preferred_model_capabilities: [...decision.preferred_model_capabilities], next_actions: [...decision.next_actions] });
  result.secret_material = "never_returned";
  return result;
}

export function inferAutonomousTaskDecision(args: {
  intent: AutonomousTaskIntent;
  lens: AutonomousDomainTaskLens;
  policy: AutonomousDomainPolicy;
  requiredModelCapabilities: readonly string[];
}): AutonomousTaskDecision {
  if (!args.intent || args.intent.schema !== "bioprism-autonomous-task-intent/0.1" || !args.lens || args.lens.schema !== "bioprism-autonomous-domain-task-lens/0.1" || !args.policy || args.policy.schema !== "bioprism-autonomous-domain-policy/0.1") throw new ArgumentError("task decision requires a valid intent, lens, and policy");
  if (args.intent.domain !== args.lens.domain || args.intent.domain !== args.policy.domain) throw new ArgumentError("task decision intent, lens, and policy domains must agree");
  const required = items("task decision requiredModelCapabilities", args.requiredModelCapabilities);
  if (!required.length) throw new ArgumentError("task decision requires at least one model capability");
  const action = args.intent.action_mode;
  const path = args.intent.domain === "cross_domain" || action === "coordinate" || action === "synthesize"
    ? "cross_domain"
    : action === "create" || action === "modify"
      ? "workflow"
      : action === "plan"
        ? "planning"
        : args.policy.evidence_mode === "required_before_provider" || ["observe", "investigate", "analyze", "compare", "evaluate"].includes(action)
          ? "evidence_first"
          : "provider";
  const preferred = unique([
    ...args.lens.model_capability_hints,
    ...(path === "workflow" ? ["structured_output"] : []),
    ...(path === "cross_domain" ? ["coordination", "structured_output"] : []),
    ...(["analyze", "compare", "evaluate", "synthesize"].includes(action) ? ["reasoning"] : []),
  ]);
  const approvals: string[] = ["provider_call"];
  const review: string[] = [];
  const blocked: string[] = [];
  if (args.policy.evidence_mode === "required_before_provider") { approvals.push("evidence_dispatch"); review.push("evidence_required_before_provider"); }
  if (["workflow", "planning", "cross_domain"].includes(path) && args.policy.plan_acceptance_required) { approvals.push("plan_acceptance"); review.push("plan_acceptance_required"); }
  if (args.intent.requested_effect === "external_effect") {
    if (args.policy.effect_mode === "forbidden") blocked.push("requested_effect_forbidden_by_domain_policy");
    else { approvals.push("effect_approval"); review.push("external_effect_requires_explicit_approval"); }
  } else if (args.intent.requested_effect === "local_change" && args.policy.effect_mode === "approval_gated") { approvals.push("effect_approval"); review.push("local_change_requires_explicit_approval"); }
  if (args.policy.evaluator_required) approvals.push("evaluator_settlement");
  for (const flag of args.intent.ambiguity_flags) review.push(`intent:${flag}`);
  if (args.intent.domain === "cross_domain" || action === "coordinate" || action === "synthesize") review.push("specialist_boundaries_require_review");
  if (args.intent.risk_signals.length) review.push("domain_risk_signals_require_review");
  const approvalRequirements = unique(approvals);
  const reviewReasons = unique(review);
  const blockingReasons = unique(blocked);
  const posture = blockingReasons.length ? "blocked" : reviewReasons.length ? "review_required" : "admitted";
  const nextActions = blockingReasons.length
    ? ["stop_before_provider_dispatch", "resolve_domain_policy_conflict"]
    : reviewReasons.length
      ? ["review_task_intent_and_decision", "satisfy_required_approval_gates", ...(approvalRequirements.includes("evidence_dispatch") ? ["acquire_and_review_required_evidence"] : [])]
      : ["select_model", "request_provider_call_approval"];
  if (approvalRequirements.includes("evaluator_settlement")) nextActions.push("settle_explicit_evaluator_feedback_after_run");
  const descriptor: DecisionDescriptor = {
    schema: AUTONOMOUS_TASK_DECISION_SCHEMA,
    decision_version: AUTONOMOUS_TASK_DECISION_VERSION,
    domain: args.intent.domain,
    workflow_id: args.intent.workflow_id,
    task_digest: args.intent.task_digest,
    intent_id: args.intent.intent_id,
    intent_digest: args.intent.intent_digest,
    lens_digest: args.lens.lens_digest,
    policy_digest: args.policy.policy_digest,
    decision_id: `${args.intent.intent_id}:${posture}:${path}`,
    posture,
    recommended_path: path,
    requested_effect: args.intent.requested_effect,
    evidence_posture: args.policy.evidence_mode,
    required_model_capabilities: required,
    preferred_model_capabilities: preferred,
    approval_requirements: approvalRequirements,
    review_reasons: reviewReasons,
    blocking_reasons: blockingReasons,
    next_actions: unique(nextActions),
  };
  return Object.freeze({ ...descriptor, decision_digest: autonomousTaskDecisionDigest(descriptor), authorization: "guidance_only;provider_source_tool_and_effect_authority_remain_separate", retention: "value_only_decision_metadata;task_text_not_retained", secret_material: "never_returned" }) as AutonomousTaskDecision;
}

/**
 * Validate a persisted decision and optionally replay it against live task artifacts.
 *
 * The serialized decision is guidance metadata, never an authorization token. Structural
 * validation checks its canonical digest, bounded fields, and markers. When the original intent,
 * lens, and policy are supplied, deterministic inference is rerun and every descriptor field
 * must match before a caller crosses a provider, source, tool, evaluator, or effect boundary.
 */
export function validateAutonomousTaskDecision(
  value: AutonomousTaskDecision | unknown,
  options: {
    intent?: AutonomousTaskIntent;
    lens?: AutonomousDomainTaskLens;
    policy?: AutonomousDomainPolicy | JsonObject;
    requiredModelCapabilities?: readonly string[];
  } = {},
): AutonomousTaskDecision {
  let decision: AutonomousTaskDecision;
  if (value && typeof value === "object" && !Array.isArray(value) && (value as Partial<AutonomousTaskDecision>).schema === AUTONOMOUS_TASK_DECISION_SCHEMA) {
    const candidate = value as Record<string, unknown>;
    const allowed = new Set([
      "schema", "decision_version", "domain", "workflow_id", "task_digest", "intent_id", "intent_digest",
      "lens_digest", "policy_digest", "decision_id", "posture", "recommended_path", "requested_effect",
      "evidence_posture", "required_model_capabilities", "preferred_model_capabilities", "approval_requirements",
      "review_reasons", "blocking_reasons", "next_actions", "decision_digest", "authorization", "retention",
      "secret_material",
    ]);
    if (Object.keys(candidate).some((key) => !allowed.has(key))) throw new ArgumentError("task decision contains unsupported fields");
    if (
      candidate.authorization !== "guidance_only;provider_source_tool_and_effect_authority_remain_separate"
      || candidate.retention !== "value_only_decision_metadata;task_text_not_retained"
      || candidate.secret_material !== "never_returned"
    ) throw new ArgumentError("task decision markers are invalid");
    const asItems = (name: string): string[] => {
      const raw = candidate[name];
      if (!Array.isArray(raw)) throw new ArgumentError(`task decision ${name} must be a sequence`);
      return items(`task decision ${name}`, raw);
    };
    const domain = text("task decision domain", candidate.domain);
    if (!AUTONOMOUS_TASK_INTENT_DOMAINS.includes(domain as typeof AUTONOMOUS_TASK_INTENT_DOMAINS[number])) throw new ArgumentError("task decision domain is unsupported");
    const workflowId = text("task decision workflow_id", candidate.workflow_id);
    const decisionId = text("task decision decision_id", candidate.decision_id);
    const taskDigest = digest("task decision task_digest", candidate.task_digest);
    const intentDigest = digest("task decision intent_digest", candidate.intent_digest);
    const lensDigest = digest("task decision lens_digest", candidate.lens_digest);
    const policyDigest = digest("task decision policy_digest", candidate.policy_digest);
    const decisionDigest = digest("task decision decision_digest", candidate.decision_digest);
    if (candidate.decision_version !== AUTONOMOUS_TASK_DECISION_VERSION) throw new ArgumentError("unsupported task-decision version");
    if (!AUTONOMOUS_TASK_DECISION_POSTURES.includes(candidate.posture as typeof AUTONOMOUS_TASK_DECISION_POSTURES[number])) throw new ArgumentError("task decision posture is unsupported");
    if (!AUTONOMOUS_TASK_DECISION_PATHS.includes(candidate.recommended_path as typeof AUTONOMOUS_TASK_DECISION_PATHS[number])) throw new ArgumentError("task decision recommended_path is unsupported");
    if (!["none", "local_change", "external_effect"].includes(candidate.requested_effect as string)) throw new ArgumentError("task decision requested_effect is unsupported");
    const evidencePosture = text("task decision evidence_posture", candidate.evidence_posture);
    if (!AUTONOMOUS_TASK_DECISION_EVIDENCE_POSTURES.includes(evidencePosture as typeof AUTONOMOUS_TASK_DECISION_EVIDENCE_POSTURES[number])) throw new ArgumentError("task decision evidence_posture is unsupported");
    const descriptor: DecisionDescriptor = {
      schema: AUTONOMOUS_TASK_DECISION_SCHEMA,
      decision_version: AUTONOMOUS_TASK_DECISION_VERSION,
      domain: domain as AutonomousTaskIntent["domain"],
      workflow_id: workflowId,
      task_digest: taskDigest,
      intent_id: text("task decision intent_id", candidate.intent_id),
      intent_digest: intentDigest,
      lens_digest: lensDigest,
      policy_digest: policyDigest,
      decision_id: decisionId,
      posture: candidate.posture as AutonomousTaskDecision["posture"],
      recommended_path: candidate.recommended_path as AutonomousTaskDecision["recommended_path"],
      requested_effect: candidate.requested_effect as AutonomousTaskDecision["requested_effect"],
      evidence_posture: evidencePosture as AutonomousTaskDecision["evidence_posture"],
      required_model_capabilities: asItems("required_model_capabilities"),
      preferred_model_capabilities: asItems("preferred_model_capabilities"),
      approval_requirements: asItems("approval_requirements"),
      review_reasons: asItems("review_reasons"),
      blocking_reasons: asItems("blocking_reasons"),
      next_actions: asItems("next_actions"),
    };
    const approvals = descriptor.approval_requirements as string[];
    if (approvals.some((approval) => !AUTONOMOUS_TASK_DECISION_APPROVALS.includes(approval as typeof AUTONOMOUS_TASK_DECISION_APPROVALS[number]))) throw new ArgumentError("task decision approval_requirements contains an unsupported gate");
    if (autonomousTaskDecisionDigest(descriptor) !== decisionDigest) throw new ArgumentError("task decision digest does not match its metadata");
    decision = Object.freeze({
      ...descriptor,
      decision_digest: decisionDigest,
      authorization: "guidance_only;provider_source_tool_and_effect_authority_remain_separate",
      retention: "value_only_decision_metadata;task_text_not_retained",
      secret_material: "never_returned",
    }) as AutonomousTaskDecision;
  } else {
    throw new ArgumentError("task decision must be an object");
  }

  const supplied = [options.intent, options.lens, options.policy];
  if (supplied.some((entry) => entry !== undefined)) {
    if (!options.intent || !options.lens || !options.policy) throw new ArgumentError("task decision replay requires intent, lens, and policy together");
    const reviewedLens = validateAutonomousDomainTaskLens(options.lens);
    const reviewedIntent = validateAutonomousTaskIntent(options.intent, { lens: reviewedLens });
    const reviewedPolicy = validateAutonomousDomainPolicy(options.policy, reviewedIntent.domain);
    const replay = inferAutonomousTaskDecision({
      intent: reviewedIntent,
      lens: reviewedLens,
      policy: reviewedPolicy,
      requiredModelCapabilities: options.requiredModelCapabilities ?? decision.required_model_capabilities,
    });
    if (JSON.stringify(descriptorFor(replay)) !== JSON.stringify(descriptorFor(decision))) throw new ArgumentError("task decision does not match the supplied intent, lens, and policy");
  } else if (options.requiredModelCapabilities !== undefined) {
    throw new ArgumentError("task decision replay capabilities require intent, lens, and policy");
  }
  return decision;
}
